geometric_kernels/spaces/homogeneous_spaces.py

```diff
@@ -0,0 +1,161 @@
+        self._characters = [AveragedLieGroupCharacter(self.average_order, character) for character in
+                            G_eigenfunctions._characters]
+        self.G_torus_representative = G_eigenfunctions._torus_representative
+        self.G_difference = G_eigenfunctions._difference
+    def _compute_dimensions(self, signature):
+        raise NotImplementedError
+
+    def _torus_representative(self, X):
+        return self.G_torus_representative(X)
+        """The function maps Lie Group Element X to T -- a maximal torus of the Lie group
+        [b, n, m] ---> [b, rank, h]"""
+
+    def inverse(self, X):
+        """The function that computes inverse element in the group"""
+        raise NotImplementedError
+
+    def _difference(self, X: B.Numeric, X2: B.Numeric) -> B.Numeric:
+        """X -- [a,n,m], X2 -- [b,n,m] --> [a,b,n,n]"""
+        g = self.M.embed_manifold(X)
+        g2 = self.M.embed_manifold(X2)
+        diff = self.G_difference(g, g2)
+        return diff
+
+    def _addition_theorem(self, X: B.Numeric, X2: B.Numeric, **parameters) -> B.Numeric:
+        diff = self._difference(X, X2)
+        diff_h = self.G_difference(diff, self.H_samples)
+        torus_repr_diff = self.G_torus_representative(diff_h)
+        values = [
+            degree * chi(torus_repr_diff)[..., None]  # [N1, N2, 1]
+            for chi, degree in zip(self._characters, self._dimensions)
+        ]
+        return B.concat(*values, axis=-1)  # [N, N2, L]
+
+    def _addition_theorem_diag(self, X: B.Numeric, **parameters) -> B.Numeric:
+        """
+        Returns the sum of eigenfunctions on a level for which we have a simplified expression
+        :param X: [N, D]
+        :param parameters: any additional parameters
+        :return: Evaluate the sum of eigenfunctions on each level. Returns
+            a value for each level [N, L]
+        """
+        g = self.M.embed_manifold(X)
+        g_h = self._difference(g, self.H_samples)
+        torus_repr_X = self._torus_representative(g_h)
+        values = [
+            degree * chi(torus_repr_X)  # [N, 1]
+            for chi, degree in zip(self._characters, self._dimensions)
+        ]
+        return B.concat(*values, axis=1)  # [N, L]
+
+    @property
+    def num_levels(self) -> int:
+            """Number of levels, L"""
+            return self._num_levels
+
+    @property
+    def num_eigenfunctions(self) -> int:
+        """Number of eigenfunctions, M"""
+        return self._num_eigenfunctions
+
+    @property
+    def num_eigenfunctions_per_level(self) -> B.Numeric:
+        """Number of eigenfunctions per level"""
+        raise NotImplementedError
+
+    def __call__(self, X: B.Numeric):
+        gammas = self._torus_representative(X)
+        res = []
+        for chi in self._characters:
+            res.append(chi(gammas))
+        res = B.stack(res, axis=1)
+        return res
+
+
+class AveragedLieGroupCharacter(abc.ABC):
```

Comment:
Add docstring for the class and its methods

Reply:
Maybe it makes more sense to generate random shifts in this class, not just accept them as an argument of `__call__`?

Reply:
This might be a nice idea but looks like it would disrupt too much of the existing pipeline... I suggest leaving it as it is for now not to delay this PR any further